### PR TITLE
0322 0198 0058

### PR DESCRIPTION
**Pull request by tushar8049**

### CODE REVIEW

**Comment by tushar-gonawala** [comment]
@pradhan1234

File: 0035. Search Insert Position/Solution.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+         * Standard Example of Binary Search with handling the edge cases.
+         */
+        while(i <= j){
+            int mid = i + (j-i)/2;
+            /** If the target is found the return the index*/
+            if (nums[mid] == target){
+                return mid;
+            }
+
+            /** If the terget is to be inserted at the end of the array or
```
Comment (pradhan1234) [inline 0035. Search Insert Position/Solution.java]: typo here

File: 0035. Search Insert Position/Solution.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+             * Target is greater than mid and less than mid + 1 */
+            if(target > nums[mid] && mid + 1 == nums.length){
+                return nums.length;
+            }
+            else if(target > nums[mid] && target < nums[mid+1]){
+                return mid + 1;
+            }
+
+            /** If the terget is to be inserted at the begining of the array or
```
Comment (pradhan1234) [inline 0035. Search Insert Position/Solution.java:38]: typo

File: 0518. Coin Change 2/Solution.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+class Solution {
+    public static void main(String[] args){
+        int[] arr = new int[]{1,2,5};
+        System.out.println("Answer: "+change(5, arr));
+    }
+
+    public static int change(int amount, int[] coins) {
+        /**
+         * Memoization Data Structure used is an 1-D array that willbe used to store the previous states
```
Comment (pradhan1234) [inline 0518. Coin Change 2/Solution.java]: add space

**Review by pradhan1234 — CHANGES_REQUESTED** [review]

File: 0035. Search Insert Position/Solution.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+ * If not, return the index where it would be if it were inserted in order.
+ *
+ * You may assume no duplicates in the array.
+ */
+
+
+/**
+ * Time Complexity: O(log n)
+ * Space Complexity: O(1)
+ */
```
Comment (pradhan1234) [inline 0035. Search Insert Position/Solution.java:12]: nice

File: 0035. Search Insert Position/Solution.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+            if(nums[mid] > target){
+                j = mid-1;
+            }
+            else{
+                i = mid + 1;
+            }
+        }
+        return i;
+    }
+}
```
Comment (pradhan1234) [inline 0035. Search Insert Position/Solution.java]: looks good

File: 0058. Length of Last Word/Solution.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+ * return the length of last word in the string.
+ *
+ * If the last word does not exist, return 0.
+ *
+ * Note: A word is defined as a character sequence consists of non-space characters only.
+ */
+
+/**
+ * Time Complexity: O(n)
+ * Space Complexity: O()
```
Comment (pradhan1234) [inline 0058. Length of Last Word/Solution.java]: O(void)???

File: 0058. Length of Last Word/Solution.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+        }
+        int i = len - 1;
+
+        /** Start counting the final length of the word. */
+        while(i >= 0 && s.charAt(i)!=' '){
+            i--;
+        }
+        return len - i - 1 ;
+    }
+}
```
Comment (pradhan1234) [inline 0058. Length of Last Word/Solution.java]: Does the context of question allow us to modify the string, well that might lead us using extra space, but put at use some good functions like trim() etc.
anyways, this is clear :)

File: 0198. House Robber/Solution.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+            return nums[0];
+        }
+        if(len == 2){
+            return Math.max(nums[0], nums[1]);
+        }
+
+        int first = 0;
+        int second = 0;
+        int ans = 0;
+
```
Comment (pradhan1234) [inline 0198. House Robber/Solution.java:40]: nice that you have deduced and not created an array to hold computation.
for me, that came at a later stage,

File: 0198. House Robber/Solution.java [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+        /** The Decision of choosing the house to rob depends upon the previous two houses and maximizing the money.*/
+        for(int i = 0; i < len; i++){
```
Comment (pradhan1234) [inline 0198. House Robber/Solution.java:42]: one line comment about what exactly are we optimizing, something like
f(k) = max(f(k – 2) + A[k], f(k – 1))
for me its clear anyways

File: 0322. Coin Change/Solution.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
+/**
+ * You are given coins of different denominations and a total amount of money amount.
+ * Write a function to compute the fewest number of coins that you need to make up that amount.
+ * If that amount of money cannot be made up by any combination of the coins, return -1.
+ */
+
+/**
+ * Time Complexity: O(nm)    n is the amount and m is the number of coins
+ * Space Complexity: O(n)
```
Comment (pradhan1234) [inline 0322. Coin Change/Solution.java:9]: nice that you reduced the space complexity, I intuitively prefer the matrix filling which gives O(m*n)

File: 0322. Coin Change/Solution.java [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+         */
+        for(int i = 1; i <= amount; i++){
+            memo[i] = amount+1;
+            for(int coin: coins) {
+                if(coin <= i)
+                    memo[i] = Math.min(memo[i], memo[i - coin] + 1);
+            }
+        }
+        /** Handling the case where the amount can not be generated using the given coins. */
+        if(memo[amount] > amount){
```
Comment (pradhan1234) [inline 0322. Coin Change/Solution.java:36]: good use of sentinels, could use something like MAX = AMOUNT + 1, and use that constant, would make it more easier to interpret, but its perfect :)

**Review by pradhan1234 — CHANGES_REQUESTED** [review]

File: 0058. Length of Last Word/Solution.java (hunk 7 above)
Comment (tushar8049) [reply 0058. Length of Last Word/Solution.java]: I am not actually manipulating the string just messing around with the variable _len_ , so ultimately the space remains constant.

File: 0322. Coin Change/Solution.java (hunk 10 above)
Comment (tushar8049) [reply 0322. Coin Change/Solution.java:9]: That would be the intuition but while coding you realize that you don't quite need a matrix.

Comment (pradhan1234) [reply 0322. Coin Change/Solution.java:9]: pro @tushar8049

**Review by pradhan1234 — APPROVED** [review]

**Pull request merged**